src/zmalloc.c

```diff
@@ -67,10 +67,34 @@ void zlibc_free(void *ptr) {
 #define dallocx(ptr,flags) je_dallocx(ptr,flags)
 #endif
 
-#define update_zmalloc_stat_alloc(__n) atomicIncr(used_memory,(__n))
-#define update_zmalloc_stat_free(__n) atomicDecr(used_memory,(__n))
+#define MAX_THREADS 16 /* Keep it a power of 2 so we can use '&' instead of '%'. */
+#define THREAD_MASK (MAX_THREADS - 1)
+
+typedef struct used_memory_entry {
+    redisAtomic long long used_memory;
+    char padding[CACHE_LINE_SIZE - sizeof(long)];
+} used_memory_entry;
+
+static __attribute__((aligned(CACHE_LINE_SIZE))) used_memory_entry used_memory[MAX_THREADS];
+static redisAtomic size_t num_active_threads = 0;
+static __thread long my_thread_index = -1;
+
+static void init_my_thread_index(void) {
+    if (my_thread_index == -1) {
+        atomicGetIncr(num_active_threads, my_thread_index, 1);
+        my_thread_index &= THREAD_MASK;
+    }
+}
 
-static redisAtomic size_t used_memory = 0;
+static void update_zmalloc_stat_alloc(long num) {
```

Comment:
```suggestion
static void update_zmalloc_stat_alloc(long long num) {
```
This should be long long at least right? Otherwise, it may not fit into it if long is 4 bytes. e.g. on macOS